x/pki/keeper/msg_server_remove_noc_x_509_ica_cert.go

```diff
@@ -0,0 +1,99 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	pkitypes "github.com/zigbee-alliance/distributed-compliance-ledger/types/pki"
+	dclauthtypes "github.com/zigbee-alliance/distributed-compliance-ledger/x/dclauth/types"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/pki/types"
+)
+
+func (k msgServer) RemoveNocX509IcaCert(goCtx context.Context, msg *types.MsgRemoveNocX509IcaCert) (*types.MsgRemoveNocX509IcaCertResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	signerAddr, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		return nil, pkitypes.NewErrInvalidAddress(err)
+	}
+
+	// check if signer has vendor role
+	if !k.dclauthKeeper.HasRole(ctx, signerAddr, dclauthtypes.Vendor) {
+		return nil, pkitypes.NewErrUnauthorizedRole("MsgRemoveNocX509IcaCert", dclauthtypes.Vendor)
+	}
+
+	signerAccount, _ := k.dclauthKeeper.GetAccountO(ctx, signerAddr)
+	accountVid := signerAccount.VendorID
+
+	icaCerts, foundActive := k.GetNocIcaCertificatesBySubjectAndSKID(ctx, accountVid, msg.Subject, msg.SubjectKeyId)
+	revCerts, foundRevoked := k.GetRevokedCertificates(ctx, msg.Subject, msg.SubjectKeyId)
+	certificates := icaCerts.Certs
+	certificates = append(certificates, revCerts.Certs...)
+	if len(certificates) == 0 {
+		return nil, pkitypes.NewErrCertificateDoesNotExist(msg.Subject, msg.SubjectKeyId)
+	}
+
+	// Existing certificate must be NOC certificate
+	if !certificates[0].IsNoc {
+		return nil, pkitypes.NewErrProvidedNocCertButExistingNotNoc(msg.Subject, msg.SubjectKeyId)
+	}
+
+	if err = k.EnsureVidMatches(ctx, certificates[0].Owner, msg.Signer); err != nil {
+		return nil, err
+	}
+
+	certID := types.CertificateIdentifier{
+		Subject:      msg.Subject,
+		SubjectKeyId: msg.SubjectKeyId,
+	}
+
+	if msg.SerialNumber != "" {
+		certBySerialNumber, found := findCertificate(msg.SerialNumber, &certificates)
+		if !found {
+			return nil, pkitypes.NewErrCertificateBySerialNumberDoesNotExist(msg.Subject, msg.SubjectKeyId, msg.SerialNumber)
+		}
+
+		// remove from subject with serialNumber map
+		k.RemoveUniqueCertificate(ctx, certBySerialNumber.Issuer, certBySerialNumber.SerialNumber)
+
+		if foundActive {
+			// Remove from Approved lists
+			aprCerts, _ := k.GetApprovedCertificates(ctx, msg.Subject, msg.SubjectKeyId)
+			removeCertFromList(certBySerialNumber.Issuer, certBySerialNumber.SerialNumber, &aprCerts.Certs)
+			k._removeApprovedX509Cert(ctx, certID, &aprCerts, msg.SerialNumber)
```

Comment:
Maybe it would be better to move `_removeApprovedX509Cert` from `msg_server_remove_x_509_cert` file to a more generic file